src/include/migraphx/generic_float.hpp

```diff
@@ -0,0 +1,435 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2024 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <migraphx/config.hpp>
+#include <migraphx/bit_cast.hpp>
+#include <algorithm>
+#include <limits>
+#include <iostream>
+#include <tuple>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+template <unsigned int N>
+constexpr unsigned int all_ones() noexcept
+{
+    return (1u << N) - 1u;
+}
+
+template <typename T>
+constexpr int countl_zero(T value)
+{
+    unsigned int r = 0;
+    for(; value != 0u; value >>= 1u)
+        r++;
+    return 8 * sizeof(value) - r;
+}
+
+struct float32_parts
+{
+    unsigned int mantissa : 23;
+    unsigned int exponent : 8;
+    unsigned int sign : 1;
+
+    static constexpr unsigned int mantissa_width() { return 23; }
+
+    static constexpr unsigned int max_exponent() { return all_ones<8>(); }
+
+    static constexpr int exponent_bias() { return all_ones<7>(); }
+
+    constexpr float to_float() const noexcept { return migraphx::bit_cast<float>(*this); }
+};
+
+constexpr float32_parts get_parts(float f) { return migraphx::bit_cast<float32_parts>(f); }
+
+#pragma pack(push, 1)
+template <unsigned int MantissaSize, unsigned int ExponentSize, unsigned int Flags = 0>
+struct alignas(1) __attribute__((may_alias)) generic_float
```

Comment:
You need the packed attribute for gcc/clang, but this probably breaks windows though. Instead of using macros, doing `[[gnu::packed, gnu::may_alias]]` may work instead.